pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -259,6 +261,20 @@ func (configgen *ConfigGeneratorImpl) buildGatewayHTTPRouteConfig(node *model.Pr
 				continue
 			}
 
+			// Make sure we can obtain services which are visible to this virtualService.
+			nameToServiceMap, exist := namespaceAndHostname[virtualService.Namespace]
+			if !exist {
+				services := push.Services(&model.Proxy{
+					ConfigNamespace: virtualService.Namespace,
+				})
+
+				nameToServiceMap = map[host.Name]*model.Service{}
+				for _, service := range services {
+					nameToServiceMap[service.Hostname] = service
```

Comment:
In some case if there are two same services from different namespaces exported to the gateway, the one within the vs' namespace takes precedence.

Reply:
Yes. Since the services which has the same namespace as the vs always at the top of the list, we only need to pick the first service.